File: clojure/bob/src/bob.clj
(ns bob)

(defn- question? [prompt]
  (= \? (last prompt)))

(defn- yell? [prompt]
  (and (= prompt (clojure.string/upper-case prompt))
       (some #(Character/isLetter %) prompt)))

(defn- nothing? [prompt]
  (clojure.string/blank? prompt))

(defn response-for
  [prompt]
  (cond (nothing? prompt) "Fine. Be that way!"
        (yell? prompt) "Whoa, chill out!"
        (question? prompt) "Sure."
        :else "Whatever."))

File: scala/bob/src/main/scala/Bob.scala


class Bob {
  def hey(prompt: String) = {
    val letters = prompt.filter(_.isLetter)
    if (!letters.isEmpty && letters.forall(_.isUpper))
      "Whoa, chill out!"
    else if (prompt.endsWith("?"))
      "Sure."
    else if (prompt.filter(_.isLetterOrDigit).isEmpty)
      "Fine. Be that way!"
    else
      "Whatever."
  }
}